src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -1437,23 +1464,29 @@ void HNSWIndex<DataType, DistType>::appendVector(const void *vector_data, const
         }
         maxlevel_ = element_max_level;
     }
-    *flags &= ~IN_PROCESS; // reset the IN_PROCESS flag.
+    unmarkInProcess(new_element_id);
+#ifdef BUILD_TESTS
+    num_parallel_workers--;
+#endif
+}
+
+template <typename DataType, typename DistType>
+idType HNSWIndex<DataType, DistType>::safeGetEntryPointCopy() const {
+    std::unique_lock<std::mutex> lock(entry_point_guard_);
+    return entrypoint_node_;
 }
 
 template <typename DataType, typename DistType>
 idType HNSWIndex<DataType, DistType>::searchBottomLayerEP(const void *query_data, void *timeoutCtx,
                                                           VecSimQueryResult_Code *rc) const {
     *rc = VecSim_QueryResult_OK;
 
-    std::unique_lock<std::mutex> lock(entry_point_guard_);
-    if (cur_element_count == 0) {
-        return entrypoint_node_;
-    }
-    idType curr_element = entrypoint_node_;
-    lock.unlock();
+    idType curr_element = safeGetEntryPointCopy();
+    if (curr_element == HNSW_INVALID_ID)
+        return curr_element; // index is empty.
 
     DistType cur_dist = this->dist_func(query_data, getDataByInternalId(curr_element), this->dim);
-    for (size_t level = maxlevel_; level > 0 && curr_element != HNSW_INVALID_ID; level--) {
```

Comment:
Why did you remove this condition? `curr_element` can become `HNSW_INVALID_ID` if a timeout occurred in `greedySearchLevel<true>`